qcore/asserts.py

```diff
@@ -353,6 +362,51 @@ def __exit__(self, exc_type, exc_val, exc_tb):
         raise AssertionError(message)
 
 
+class AssertRaisesInstance(object):
+    """With-context that asserts that the code within the context raises a specified exception
+    instance."""
+
+    def __init__(self, *expected_exception_instances, **kwargs):
+        assert (
+            len(expected_exception_instances) >= 1
+        ), "You must specify an exception instance when using AssertRaisesInstance"
+        self.expected_exception_instances = expected_exception_instances
+        self.extra = kwargs.pop("extra", None)
+        assert_eq({}, kwargs)
+
+    def __enter__(self):
+        self.ctx = AssertRaises(
+            *[type(exception_instance) for exception_instance in self.expected_exception_instances],
+            extra=self.extra)
+        self.assert_raises = self.ctx.__enter__()
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        ret = self.ctx.__exit__(exc_type, exc_val, exc_tb)
+        if ret:
+            expected_args = []
+            for exception_instance in self.expected_exception_instances:
+                if type(exception_instance) == exc_type:
+                    if exception_instance.args == self.assert_raises.expected_exception_found.args:
+                        return True
+                    else:
+                        expected_args.append(exception_instance.args[0])
```

Comment:
Why not just use `str(exception_instance)`? Other args than the first one may be significant.

Reply:
Ok yes good point, I've updated this to show the string representation of the exception instance(s) instead.